TSRM/tsrm_win32.c

```diff
@@ -662,6 +665,19 @@ TSRM_API int shmget(key_t key, size_t size, int flags)
 		}
 	}
 
+	if (key == IPC_PRIVATE) {
+		/* This should call shm_get with a brand new key id that isn't used yet. See https://man7.org/linux/man-pages/man2/shmget.2.html
+		 * Because shmop_open can be used in userland to attach to shared memory segments, use high positive numbers to avoid accidentally conflicting with userland. */
+		key = (php_rand() & 0x3fffffff) + 0x40000000;
+		for (shm_pair *ptr = TWG(shm); ptr < (TWG(shm) + TWG(shm_size)); ptr++) {
+			if (ptr->descriptor && ptr->descriptor->shm_perm.key == key) {
+				key = (php_rand() & 0x3fffffff) + 0x40000000;
+				ptr = TWG(shm);
+				continue;
+			}
+		}
+	}
```

Comment:
I'm not sure if this makes much sense (I'm more concerned that another process/thread may use a colliding key), but it doesn't hurt.

Reply:
Thread safety on Windows ZTS for this structure seems like a pre-existing issue. Maybe create a ticket to add a mutex for the entire sections of shm_get in another PR?

Reply:
`TWG(shm)` is a per-thread local variable (TLS), so there is no thread-safety issue. My point was that other processes (or threads) may already have allocated a certain key, and that loop can't detect that.

Reply:
> So that's why apcu can't share ids across threads with the windows polyfill for shmget

That shouldn't be the reason (and from what I can tell, there shouldn't be APCu issues for threaded SAPIs). The SHM segments can even be shared between unrelated processes with the current implementation on Windows.

shmop1.php
````php
<?php

$shmop = shmop_open(12345, 'c', 0777, 1024);
shmop_write($shmop, "test", 0);
sleep(10);
````

shmop2.php
````php
<?php

$shm2 = shmop_open(12345, 'a', 0777, 1024);
$read = shmop_read($shm2, 0, 4);
var_dump($read);
````

````
> php-master shmop1.php
> php-8.2 shmop2.php
string(4) "test"
````

> I'm not quite familiar with how forking in non-threaded(cli, fpm, etc) would work on windows

There is no `fork(2)` on Windows (the Cygwin people tried to come up with an emulation, but gave up long ago). This is the main reason why there is no PHP-FPM on Windows (and we're working around that problem for a long time).